### ISSUE TITLE
Nullness Checker accepts a cast from T to T that throws a NullPointerException

### ISSUE THREAD

**Issue by aosen-xiong** [eisop/checker-framework#2011]
The Nullness Checker issues no diagnostic for the program below, which throws a
`NullPointerException` at run time.

```java
import org.checkerframework.checker.nullness.qual.NonNull;
import org.checkerframework.checker.nullness.qual.Nullable;

class Box<T extends @Nullable Object> {
    T launder(@Nullable T p) {
        T x = (T) p;
        return x;
    }
}

public class Npe {
    public static void main(String[] args) {
        Box<@NonNull String> b = new Box<>();
        @NonNull String s = b.launder(null);
        System.out.println(s.length());
    }
}
```

```
$ javac -processor org.checkerframework.checker.nullness.NullnessChecker Npe.java
$ java Npe
Exception in thread "main" java.lang.NullPointerException:
    Cannot invoke "String.length()" because "<local2>" is null
```

Expected: a diagnostic on the cast or on the call.

Actual: no diagnostic; `NullPointerException` at run time.

Reproduced on master at 352e57aedb.

**Comment by wmdietl** [comment]
Odd, the local variable should be refined to `@Nullable T` and then the return should be incompatible.

Would you expect an error at the cast or the return?

Does checkCastElementType make a difference?
https://github.com/eisop/checker-framework/blob/352e57aedbcf95f88bd5f9df40ecb836bff7be6c/framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java#L202

**Comment by aosen-xiong** [comment]
Enable the flag get `warning: [cast.unsafe]` on `T x = (T) p;`.
